File: 11/python/solve.py
import itertools

with open("../input.txt") as h:
    board = [[int(c) for c in line.strip()] for line in h.readlines()]

def add_one():
    for i, row in enumerate(board):
        for j, c in enumerate(row):
            board[i][j] = (c + 1) % 10

def flash_one(i, j):
    if i < 0 or j < 0 or i >= len(board) or j >= len(board[i]):
        return

    if board[i][j] > 0:
        board[i][j] = (board[i][j] + 1) % 10

def print_board():
    for row in board:
        print(''.join(str(c) for c in row))
    print()

DT = [(-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 0), (0, 1), (1, -1), (1, 0), (1, 1)]
def flash(flashed=None):
    if flashed == None:
        previous = -1
        flashed = set()
    else:
        previous = len(flashed)

    for i, row in enumerate(board):
        for j, c in enumerate(row):
            if not c == 0:
                continue

            if (i, j) in flashed:
                continue

            flashed.add((i, j))

            for di, dj in DT:
                flash_one(i + di, j + dj)

    if previous < len(flashed):
        return flash(flashed)
    else:
        return len(flashed)

t1 = 0
t2 = None
for step in itertools.count(start=1):
    add_one()
    flash_count = flash()

    if step <= 100:
        t1 += flash_count

    if t2 == None and flash_count >= len(board) * len(board[0]):
        t2 = step

        if step > 100:
            break

print("Task 1:", t1)
print("Task 2:", t2)

File: 20/ruby/solve.rb
require 'set'
require 'matrix'

data = File.open("../input.txt").readlines.map(&:chomp).map{|line| line.chars.map{|c| if c == "#" then 1 else 0 end}}
$algorithm = data[0]
$deltas = [[-1,-1],[-1,0],[-1,1],[0,-1],[0,0],[0,1],[1,-1],[1,0],[1,1]].map{|a| Vector.elements(a)}

def out_of_bound(vector, i_min, i_max, j_min, j_max)
    vector[0] < i_min || vector[0] > i_max || vector[1] < j_min || vector[1] > j_max
end

def enhance(bound_image)
    bound, image = bound_image
    i_min, i_max = image.map{|v| v[0]}.minmax
    j_min, j_max = image.map{|v| v[1]}.minmax
    new_image = Set[]

    (i_min-1..i_max+1).each{|i| (j_min-1..j_max+1).each{|j| 
        v = Vector[i,j]

        bin = $deltas.map{|d| ((image.include? (v + d)) || (bound && out_of_bound(v + d, i_min, i_max, j_min, j_max))) ? 1 : 0}.join
        index = bin.to_i(2) 

        if $algorithm[index] == 1 then
            new_image.add(v)
        end
    }}

    new_bound = if $algorithm[0] == 1 then !bound else false end

    [new_bound, new_image]
end

def n_enhance(n, image)
    (0..n-1).each{|_| image = enhance image}

    image
end

image = [false, data[2..].map.with_index.inject(Set[]){|c, entry|
    row, i = entry
    row.each_with_index{|v,j| if v == 1 then c.add(Vector[i,j]) end}
    c
}]

print "Task 1: ", (n_enhance(2, image))[1].count, "\n"
print "Task 2: ", (n_enhance(50, image))[1].count, "\n"
